lib/matplotlib/cm.py

```diff
@@ -333,34 +324,41 @@ def set_clim(self, vmin=None, vmax=None):
 
     def set_cmap(self, cmap):
         """
-        set the colormap for luminance data
+        Set the colormap for luminance data.
 
         Parameters
         ----------
-        cmap : colormap or registered colormap name
+        cmap : `.Colormap` or str or None
         """
+        in_init = not hasattr(self, "cmap")
         cmap = get_cmap(cmap)
         self.cmap = cmap
-        self.changed()
+        if not in_init:
+            self.changed()  # Things are not set up properly yet.
 
     def set_norm(self, norm):
-        """Set the normalization instance.
+        """
+        Set the normalization instance.
 
         Parameters
         ----------
-        norm : `.Normalize`
+        norm : `.Normalize` or None
 
         Notes
         -----
         If there are any colorbars using the mappable for this norm, setting
         the norm of the mappable will reset the norm, locator, and formatters
         on the colorbar to default.
-
         """
+        if norm is not None and not isinstance(norm, colors.Normalize):
+            raise ValueError(
+                "'norm' must be an instance of 'mcolors.Normalize' or None")
+        in_init = not hasattr(self, "norm")
```

Comment:
That could easily be broken in the future when somebody edits `__init__`. Also, it's good practice to actually define the attributes in `__init__`, which this pattern relies on not to happen.

Can we do this more explicit with:
~~~
def __init__(self, norm):
    self.norm = None
    self.set_norm(norm)

def set_norm():
     was_uninitialized = self.norm is None
    ...
    if self.was_uninitialized:
        self.changed()
~~~

or maybe even more explicit with `_UNITIALIZED = object()` instead of None.

Reply:
sure, inited them to None (_UNINITIALIZED seems a bit overengineered for this)